genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java

```diff
@@ -96,104 +133,166 @@ public JobCoordinatorServiceImpl(
+                        + maxJobMemory
+                        + " MB allowed"
                 );
-                jobBuilder
-                    .withStatus(JobStatus.INIT)
-                    .withStatusMsg("Job Accepted and in initialization phase.");
-                // TODO: if this throws exception the job will never be marked failed
-                this.jobPersistenceService.createJobAndJobExecution(jobBuilder.build(), jobExecution);
-                try {
-                    log.info("Scheduling job {} for submission", jobRequest.getId());
-                    final Future<?> task = this.taskExecutor.submit(
-                        new JobLauncher(this.jobSubmitterService, jobRequest, this.registry)
+            }
+
+            synchronized (this) {
+                log.info("Checking if can run job {} on this node", jobRequest.getId());
+                final int maxSystemMemory = this.jobsProperties.getMemory().getMaxSystemMemory();
+                final int usedMemory = this.jobMetricsService.getUsedMemory();
+                if (usedMemory + memory <= maxSystemMemory) {
+                    log.info(
+                        "Job {} can run on this node as only {}/{} MB are used and requested {} MB",
+                        jobId,
+                        usedMemory,
+                        maxSystemMemory,
+                        memory
                     );
+                    try {
+                        log.info("Scheduling job {} for submission", jobRequest.getId());
+                        final Future<?> task = this.taskExecutor.submit(
+                            new JobLauncher(
+                                this.jobSubmitterService,
+                                jobRequest,
+                                cluster,
+                                command,
+                                applications,
+                                memory,
+                                this.registry
+                            )
+                        );
 
-                    // Tell the system a new job has been scheduled so any actions can be taken
-                    log.info("Publishing job scheduled event for job {}", jobId);
-                    this.eventPublisher.publishEvent(new JobScheduledEvent(jobId, task, this));
-                } catch (final TaskRejectedException e) {
-                    final String errorMsg = "Unable to launch job due to exception: " + e.getMessage();
-                    this.jobPersistenceService.updateJobStatus(jobId, JobStatus.FAILED, errorMsg);
-                    throw new GenieServerException(errorMsg, e);
+                        // Tell the system a new job has been scheduled so any actions can be taken
+                        log.info("Publishing job scheduled event for job {}", jobId);
+                        this.eventPublisher.publishEvent(new JobScheduledEvent(jobId, task, memory, this));
```

Comment:
As discussed we wouldn't then have the handle to the Future returned from the scheduler that we can use to cancel the task. For now leaving as is.